org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/BreakpointManager.java

```diff
@@ -0,0 +1,172 @@
+* Copyright (c) 2017 Microsoft Corporation and others.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Eclipse Public License v1.0
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v10.html
+*
+* Contributors:
+*     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package org.eclipse.jdt.ls.debug.adapter;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+
+import org.eclipse.jdt.ls.debug.IBreakpoint;
+import org.eclipse.jdt.ls.debug.internal.Logger;
+
+public class BreakpointManager {
+    /**
+     * A collection of breakpoints registered with this manager.
+     */
+    private List<IBreakpoint> breakpoints;
+    private HashMap<String, HashMap<String, IBreakpoint>> sourceToBreakpoints;
+
+    /**
+     * Constructor.
+     */
+    public BreakpointManager() {
+        this.breakpoints = Collections.synchronizedList(new ArrayList<>(5));
+        this.sourceToBreakpoints = new HashMap<>();
+    }
+
+    /**
+     * Adds breakpoints to breakpoint manager.
+     * Deletes all breakpoints that are no longer listed.
+     * @param source
+     *              source path of breakpoints
+     * @param breakpoints
+     *              full list of breakpoints that locates in this source file
+     * @return the full breakpoint list that locates in the source file
+     */
+    public IBreakpoint[] addBreakpoints(String source, IBreakpoint[] breakpoints) {
+        return addBreakpoints(source, breakpoints, false);
+    }
+
+    /**
+     * Adds breakpoints to breakpoint manager.
+     * Deletes all breakpoints that are no longer listed.
+     * In the case of modified source, delete everything.
+     * @param source
+     *              source path of breakpoints
+     * @param breakpoints
+     *              full list of breakpoints that locates in this source file
+     * @param sourceModified
+     *              the source file are modified or not.
+     * @return the full breakpoint list that locates in the source file
+     */
+    public IBreakpoint[] addBreakpoints(String source, IBreakpoint[] breakpoints, boolean sourceModified) {
+        List<IBreakpoint> result = new ArrayList<>();
+        HashMap<String, IBreakpoint> breakpointMap = this.sourceToBreakpoints.get(source);
+        // When source file is modified, delete all previous added breakpoints.
+        if (sourceModified && breakpointMap != null) {
+            for (IBreakpoint bp : breakpointMap.values()) {
+                try {
+                    // Destroy the breakpoint on the debugee VM.
+                    bp.close();
+                } catch (Exception e) {
+                    Logger.logException("Remove breakpoint exception", e);
+                }
+                synchronized (breakpoints) {

```

Comment:
Why do we need to synchronize this operation? Set breakpoint operation can only be trigger by requests from VSCode but never the event hub.